### PR TITLE
MQTT client setup and publishing message

### PR DESCRIPTION
**Pull request by Jdallos**

### CODE REVIEW

File: android/app/build.gradle [hunk 1]
```diff
@@ -48,7 +48,7 @@ android {
         applicationId "fi.metatavu.noheva_visitor_ui"
         // You can update the following values to match your application needs.
         // For more information, see: https://docs.flutter.dev/deployment/android#reviewing-the-gradle-build-configuration.
-        minSdkVersion flutter.minSdkVersion
+        minSdkVersion 19
```
Comment (anttileppa) [inline android/app/build.gradle]: is this set to 19 on purpose? SDK 19 is really really old (see https://developer.android.com/tools/releases/platforms)

File: lib/utils/device_info.dart [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+class DeviceInfo {
+  static Future<String?> getSerialNumber() async {
+    DeviceInfoPlugin deviceInfoPlugin = DeviceInfoPlugin();
+
+    if (Platform.isAndroid) {
+      const androidId = AndroidId();
+      String? deviceId = await androidId.getId();
+
+      if (deviceId == null) {
+        SimpleLogger().warning("Device ID not found, cannot connect to MQTT.");
```
Comment (anttileppa) [inline lib/utils/device_info.dart]: Device info is not exactly related to MQTT, so this error message sounds a bit incorrect

**Review by anttileppa — CHANGES_REQUESTED** [review]

File: android/app/build.gradle (hunk 1 above)
Comment (villejuutila) [reply android/app/build.gradle]: `flutter.minSdkVersion` is even lower. It has been set due to a particular dependency (`device_info_plus` if I remember correctly) requires minimum SDK version of 19.

Would it be best to be kept as low as possible, or to correspond with initial customers Android device specifications?

File: lib/main.dart [hunk 3]
```diff
@@ -23,9 +25,35 @@ void main() async {
...
+      mqttClient.sendStatusMessage(true);
+    }
+  });
+}
+
+Future<void> _initializeMqttClient() async {
+  final String? serialNumber = await DeviceInfo.getSerialNumber();
+
+  if (serialNumber == null) {
+    SimpleLogger().warning("Device ID not found, cannot connect to MQTT.");
```
Comment (villejuutila) [inline lib/main.dart]: Device ID does sound a bit misleading as it seems to refer to a Device within Noheva which has not been created at this point yet.

However this is probably subject to change once API support for Devices is ready

File: lib/main.dart [hunk 4]
```diff
@@ -23,9 +25,35 @@ void main() async {
...
+
+    return;
+  } else {
+    SimpleLogger().info("Device ID: $serialNumber");
```
Comment (villejuutila) [inline lib/main.dart]:
```suggestion
    SimpleLogger().info("Serial number: $serialNumber");
```

File: lib/mqtt/model/status_message.dart [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import 'package:json_annotation/json_annotation.dart';
+import 'package:noheva_visitor_ui/mqtt/model/abstract_message.dart';
```
Comment (villejuutila) [inline lib/mqtt/model/status_message.dart]: Replace single quotes with double quotes. It's a shame you apparentally can't configure Dart autoformatter to do it for you.

Same applies for the whole project.

File: lib/mqtt/mqtt_client.dart [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+      SimpleLogger().info("MQTT Client already connected");
+
+      return;
+    }
+
+    if (await DeviceInfo.getSerialNumber() == null) {
+      SimpleLogger().warning("Device ID not found, cannot connect to MQTT.");
+
+      return;
+    }
```
Comment (villejuutila) [inline lib/mqtt/mqtt_client.dart]: Return value of `DeviceInfo.getSerialNumber()` is already passed to this `connect()` method as a parameter and this same check is already done in `main.dart`

File: lib/mqtt/mqtt_client.dart [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+        .startClean()
+        .withWillQos(MqttQos.atLeastOnce);
+    client.connectionMessage = connMessage;
+    try {
+      await client.connect(
+        mqttUsername,
+        mqttPassword,
+      );
+    } catch (exception) {
+      SimpleLogger().shout("Exception: $exception");
```
Comment (villejuutila) [inline lib/mqtt/mqtt_client.dart]:
```suggestion
      SimpleLogger().shout("Exception while connecting MQTT Client: $exception");
```

File: lib/mqtt/mqtt_client.dart [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+      } else {
+        SimpleLogger()
+            .warning("Didn't have listener for topic ${messages[0].topic}");
+      }
+    });
+  }
+
+  /// Handler for successful connections event.
+  ///
+  /// Will send resend status message periodically
```
Comment (villejuutila) [inline lib/mqtt/mqtt_client.dart]:
```suggestion
  /// Sends status message to status topic
```

File: lib/mqtt/mqtt_client.dart [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+      await _statusTopic,
+      createMessagePayload(jsonEncode(statusMessage)),
+    );
+  }
+
+  /// Handler for disconnection event.
+  ///
+  /// Attempts to reconnect the client.
+  void onDisconnected() {
+    SimpleLogger().info("Disconnected");
```
Comment (villejuutila) [inline lib/mqtt/mqtt_client.dart]:
```suggestion
    SimpleLogger().info("MQTT Client disconnected");
```

File: lib/mqtt/mqtt_client.dart [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+  }
+
+  /// Handler for unsubscribing from topic events.
+  void onUnsubscribed(String? topic) {
+    SimpleLogger().info("Unsubscribed from topic: $topic");
+  }
+
+  /// Disconnects MQTT Client
+  void disconnect() {
+    SimpleLogger().info("Disconnecting...");
```
Comment (villejuutila) [inline lib/mqtt/mqtt_client.dart]:
```suggestion
    SimpleLogger().info("Disconnecting MQTT Client...");
```

File: lib/mqtt/mqtt_client.dart [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+
+    return StatusMessage(
+      status: status ? "ONLINE" : "OFFLINE",
+      deviceId: deviceId,
+    );
+  }
+
+  /// Reconnects MQTT Client.
+  ///
+  /// Reports Sentry issue every 100th failure.
```
Comment (villejuutila) [inline lib/mqtt/mqtt_client.dart]:
```suggestion
  /// Reconnects MQTT Client.
```

File: pubspec.yaml [hunk 12]
```diff
@@ -23,13 +23,19 @@ dependencies:
   simple_logger: ^1.9.0+2
   noheva_api:
     path: noheva-api
+  device_info_plus: ^9.0.3
+  mqtt_client: ^10.0.0
+  typed_data: ^1.3.1
+  json_annotation: any
```
Comment (villejuutila) [inline pubspec.yaml]: Use specific version number instead of any

File: pubspec.yaml [hunk 13]
```diff
@@ -23,13 +23,19 @@ dependencies:
...
+  android_id: ^0.3.3
 
 dev_dependencies:
   flutter_test:
     sdk: flutter
   flutter_lints: ^2.0.0
   openapi_generator: ^4.0.0
   build_runner: ^2.1.8
+  json_serializable: any
```
Comment (villejuutila) [inline pubspec.yaml]: Use specific version number instead of any

**Review by villejuutila — CHANGES_REQUESTED** [review]
Requested changes

File: lib/mqtt/mqtt_client.dart (hunk 6 above)
Comment (Jdallos) [reply lib/mqtt/mqtt_client.dart]: removed

**Review by villejuutila — APPROVED** [review]

File: android/app/build.gradle (hunk 1 above)
Comment (villejuutila) [reply android/app/build.gradle]: @Jdallos bump to 28

**Pull request merged**